feat(storage): Add external storage backend integration via Sunbeam CLI

happyhackerhour opened the pull request:
# Sunbeam Storage Backend Management

## Summary
This PR introduces a comprehensive storage backend management system for Sunbeam with full unit testing coverage. It provides a clean, maintainable framework for managing Cinder storage backends, starting with Hitachi VSP support, featuring separated CLI architecture, dynamic configuration management, and 1029+ tests ensuring reliability.

## Context & Related Work
Builds on:
- **cinder-volume-hitachi charm** https://review.opendev.org/c/openstack/sunbeam-charms/+/955633 - Configures Hitachi VSP drivers
- **snap-cinder-volume* https://github.com/happyhackerhour/snap-cinder-volume/tree/feat/hitachi-backend - Backend configuration rendering  
- **This PR** - Sunbeam CLI and management framework

## Key Changes

### 1. Storage Backend Framework (`sunbeam/storage/`)
- **`base.py`** - Core service layer with `StorageBackendService` and proper CLI registration hooks
- **`registry.py`** - Dynamic backend discovery and CLI registration with plugin architecture
- **`service.py`** - Centralized Terraform state management and backend existence checking
- **`steps.py`** - Deployment workflow with comprehensive error handling
- **Exception hierarchy** for robust error handling and user feedback

### 2. Hitachi VSP Backend (`sunbeam/storage/backends/hitachi/`)
- **`backend.py`** - Core deployment and management logic with `HitachiConfig` Pydantic model
- **`cli.py`** - Dedicated CLI commands with proper separation of concerns
- **6 essential deployment fields** with comprehensive validation
- **IP/FQDN validation** with proper error handling
- **Protocol selection** (FC/iSCSI) with validation
- **Configuration filtering** logic to prevent sensitive data exposure

### 3. Dynamic Configuration Management
- **`get_backend_config()`** - Retrieve current configuration with filtering
- **CLI-based configuration management** with rich console integration
- **Sensitive value masking** for security (passwords, secrets)
- **Service layer delegation** for all Terraform state operations
- **Validation and error handling** for configuration updates


## CLI Commands

### Backend Management
```bash
# Deploy backend
sunbeam storage add <backend>                              # Interactive deployment
sunbeam storage add <backend> --name <name> [options]     # Non-interactive deployment
sunbeam storage remove <backend> <name>                   # Remove existing backend  
sunbeam storage list all                                # List all backends
```

### Configuration Management
```bash
# Configuration operations
sunbeam storage config show <backend> <name>              # View current configuration
sunbeam storage config set <backend> <name> [options]           # Update configuration
sunbeam storage config reset <backend> <name> [keys]     # Reset specific keys
sunbeam storage config options <backend>                 # List available options
```

## Example Usage

### Deploy Backend
**Interactive deployment:**
```bash
$ sunbeam storage add hitachi
SAN Management IP: 192.168.1.100
SAN Username: maintenance  
Storage Pools: pool1,pool2
Protocol: FC

✅ Successfully deployed Hitachi backend 'production-hitachi'
```

**Direct deployment with parameters:**
```bash
$ sunbeam storage add hitachi --name vsp350 \
    --san-ip=192.168.1.100 \
    --san-username=maintenance \
    --san-password=12345 \
    --hitachi-pools=pool1,pool2 \
    --protocol=FC \
    --hitachi-storage-id=123123123
```

### Manage Configuration
**View configuration:**
```bash
$ sunbeam storage config show hitachi vsp350
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┓
┃ Configuration Key  ┃ Value              ┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━┩
│ san-ip             │ 192.168.1.100      │
│ san-login          │ maintenance        │
│ hitachi-pools      │ pool1,pool2        │
│ protocol           │ FC                 │
│ san-password       │ ********           │
└────────────────────┴────────────────────┘
```

**Update configuration:**
```bash
$ sunbeam storage config set hitachi vsp350 --hitachi-copy-speed=3
✅ Configuration updated successfully
```

**List available options:**
```bash
$ sunbeam storage config options hitachi
Available configuration options for Hitachi VSP Storage:
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Option                       ┃ Description                                       ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ san-ip                       │ SAN management IP address or FQDN                │
│ san-login                    │ SAN management username                          │
│ san-password                 │ SAN management password                          │
│ hitachi-pools                │ Comma-separated list of storage pools            │
│ protocol                     │ Storage protocol (FC or iSCSI)                   │
│ hitachi-copy-speed           │ Copy speed for volume operations (1-15)          │
└──────────────────────────────┴───────────────────────────────────────────────────┘
```

wolsen reviewed (commented):
I'm currently making some inflight comments. I have not had a chance to perform an exhaustive review.

happyhackerhour committed:
feat(storage): Hitachi backend via Sunbeam CLI

happyhackerhour commented on sunbeam-python/sunbeam/storage/base.py:
```diff
@@ -0,0 +1,371 @@
+
+    def _get_backend_type(self, app_name: str) -> str:
+        """Determine backend type from application name."""
+        if "hitachi" in app_name:
```
you are rigth. will change it to use the charm name to determine the backend type

happyhackerhour replied:
done

gboutry commented:
My main comment was lost with that whole shebang of pending review...

This is overall a nice work, and thank you very much for your contribution.

Can you ensure this is inline with the rest of the code base?
Can you reduce the use of emojis?
Can you format the code? There are a lot of trailing whitespaces.

happyhackerhour commented on sunbeam-python/tests/unit/sunbeam/storage/test_storage_simple.py:
```diff
@@ -0,0 +1,303 @@
+
+                for charm_name, app_name, expected in test_cases:
+                    with self.subTest(charm_name=charm_name, app_name=app_name):
+                        result = service._get_backend_type_from_charm(
```
uploading updated unit testing and pep8 formating compliant

happyhackerhour committed:
fix: tox -e fmt, pep8, mypy

reimplement Unit testing to conform to new signatures and changes to use terraform

happyhackerhour commented on sunbeam-python/sunbeam/commands/storage.py:
```diff
@@ -0,0 +1,125 @@
+    commands dynamically based on available backends.
+    """
+    try:
+        StorageBackendRegistry().register_cli_commands(storage, deployment)
```
done

happyhackerhour committed:
Fix storage backend removal regression and address reviewer comments

- Fix partial removal regression: write updated Terraform variables before apply
- Remove global registry instance, follow Provider pattern for CLI registration
- Add Juju application name validation to prevent invalid backend names
- Implement atomic removal operations with proper rollback on failure
- Add cleanup-config command to resolve invalid backend configurations
- Improve error handling and logging throughout storage backend system
- Update deprecated Pydantic __fields__ usage for v2 compatibility
- Consolidate imports and reduce emoji usage in console output

happyhackerhour committed:
refactor: clean up Hitachi backend CLI and eliminate code duplication

- Move CLI code to dedicated module (hitachi/cli.py) with proper separation
- Remove duplicate backend_exists() methods, consolidate in service layer
- Eliminate unused service methods (set/reset_backend_config) and tests
- Refactor duplicate config filtering logic into reusable internal function
- Clean up storage backend __init__.py files to match provider pattern
- Update all CLI and backend code to use service layer properly
- Fix missing abstract method implementations in test mocks
- Update test mocking to reflect architectural changes

All tests pass (1029/1029) with clean, maintainable, well-structured codebase.

Signed-off-by: Hugo Vinicius Garcia Razera <root@happyhackerhour.net>

happyhackerhour commented on sunbeam-python/sunbeam/storage/registry.py:
```diff
@@ -0,0 +1,498 @@
+        console.print(table)
+
+# Global registry instance
+storage_backend_registry = StorageBackendRegistry()
```
Hi Guillaume, removed the global registry and followed the pattern setup by the provider module

happyhackerhour commented on sunbeam-python/sunbeam/storage/registry.py:
```diff
@@ -0,0 +1,498 @@
+        self._register_list_commands(storage_group, deployment)
+        self._register_config_commands(storage_group, deployment)
+
+    def _register_add_commands(self, storage_group: click.Group, deployment: Deployment) -> None:
```
Done

happyhackerhour committed:
- Retry apply and destroy plans on terraform locks
- Shows friendly error messege on state locks
- Verify juju login status, before add and remove.

Signed-off-by: Hugo Vinicius Garcia Razera <root@happyhackerhour.net>

happyhackerhour committed:
- Refactor common Backend Methods into Base Class
- Implement PureStorage Backend
- separate terraforms-key for each backend

happyhackerhour committed:
- Implementaion of Dell Storage Center Backend
- fix show config options for pure storage

happyhackerhour committed:
Merge branch 'canonical:main' into feat/storage-cli

gboutry commented:
Closing this pr as it has been implemented by #609 based on this branch.
